File: bots/whatsapp-bot.js
/**
 * WhatsApp Bot Handler
 * Handles incoming WhatsApp messages via Twilio
 * Processes claims and returns verification results
 */

require('dotenv').config();
const express = require('express');
const twilio = require('twilio');

// Import modules
const { analyzeMessage } = require('../modules/claimExtractor');
const { generateClaimHash } = require('../modules/hashGenerator');
const { submitClaimToBackend, getClaimResult, claimExists } = require('../modules/backendClient');
const { 
  formatWhatsAppReport, 
  formatErrorMessage, 
  formatWelcomeMessage,
  formatHelpMessage 
} = require('../utils/formatter');

// Initialize Express app
const app = express();
app.use(express.urlencoded({ extended: false }));
app.use(express.json());

// Twilio configuration
const accountSid = process.env.TWILIO_ACCOUNT_SID;
const authToken = process.env.TWILIO_AUTH_TOKEN;
const twilioClient = twilio(accountSid, authToken);

// Validate Twilio webhook (optional but recommended for production)
const validateTwilioRequest = (req, res, next) => {
  if (process.env.NODE_ENV === 'production') {
    const twilioSignature = req.headers['x-twilio-signature'];
    const url = `${req.protocol}://${req.get('host')}${req.originalUrl}`;
    
    if (!twilio.validateRequest(authToken, twilioSignature, url, req.body)) {
      return res.status(403).send('Forbidden');
    }
  }
  next();
};

/**
 * Main message handler
 * Process incoming WhatsApp messages
 */
async function handleWhatsAppMessage(messageText, fromNumber) {
  try {
    console.log(`[WhatsApp] Message from ${fromNumber}: ${messageText}`);

    // Handle commands
    const lowerText = messageText.toLowerCase().trim();
    
    if (lowerText === 'start' || lowerText === 'hi' || lowerText === 'hello') {
      return formatWelcomeMessage();
    }
    
    if (lowerText === 'help') {
      return formatHelpMessage();
    }

    // Analyze the message
    console.log('[WhatsApp] Analyzing message...');
    const analysis = await analyzeMessage(messageText);
    
    if (!analysis.claims || analysis.claims.length === 0) {
      return formatErrorMessage('invalid_message');
    }

    // Generate claim hash from first claim
    const claimText = analysis.claims[0];
    const claimHash = generateClaimHash(claimText);
    console.log(`[WhatsApp] Generated hash: ${claimHash}`);

    // Check if claim exists in blockchain
    const exists = await claimExists(claimHash);
    
    if (!exists) {
      // Submit new claim to backend
      console.log('[WhatsApp] Submitting new claim to blockchain...');
      await submitClaimToBackend(claimHash, claimText);
    }

    // Get current result from blockchain
    console.log('[WhatsApp] Fetching blockchain result...');
    const blockchainResult = await getClaimResult(claimHash);

    if (!blockchainResult.success) {
      return formatErrorMessage('api_down');
    }

    // Format and return response
    const report = formatWhatsAppReport(analysis, blockchainResult.data || blockchainResult);
    console.log('[WhatsApp] Report generated successfully');
    
    return report;

  } catch (error) {
    console.error('[WhatsApp] Error handling message:', error);
    return formatErrorMessage('general');
  }
}

/**
 * Webhook endpoint for incoming WhatsApp messages
 * Twilio sends POST requests here
 */
app.post('/webhook/whatsapp', validateTwilioRequest, async (req, res) => {
  try {
    const incomingMessage = req.body.Body || '';
    const fromNumber = req.body.From || '';
    
    console.log(`[Webhook] Received message: "${incomingMessage}" from ${fromNumber}`);

    // Process message
    const responseMessage = await handleWhatsAppMessage(incomingMessage, fromNumber);

    // Send response via Twilio
    const twiml = new twilio.twiml.MessagingResponse();
    twiml.message(responseMessage);

    res.type('text/xml');
    res.send(twiml.toString());

  } catch (error) {
    console.error('[Webhook] Error:', error);
    
    const twiml = new twilio.twiml.MessagingResponse();
    twiml.message(formatErrorMessage('general'));
    
    res.type('text/xml');
    res.send(twiml.toString());
  }
});

/**
 * Health check endpoint
 */
app.get('/health', (req, res) => {
  res.json({ 
    status: 'ok', 
    service: 'WhatsApp Bot',
    timestamp: new Date().toISOString()
  });
});

/**
 * Send outbound WhatsApp message (for testing or notifications)
 */
async function sendWhatsAppMessage(toNumber, message) {
  try {
    const twilioWhatsAppNumber = process.env.TWILIO_WHATSAPP_NUMBER;
    
    const sentMessage = await twilioClient.messages.create({
      from: `whatsapp:${twilioWhatsAppNumber}`,
      to: `whatsapp:${toNumber}`,
      body: message
    });

    console.log(`[WhatsApp] Message sent: ${sentMessage.sid}`);
    return { success: true, sid: sentMessage.sid };

  } catch (error) {
    console.error('[WhatsApp] Error sending message:', error);
    return { success: false, error: error.message };
  }
}

/**
 * Start the server
 */
const PORT = process.env.WHATSAPP_BOT_PORT || 3001;

if (require.main === module) {
  app.listen(PORT, () => {
    console.log(`✅ WhatsApp Bot server running on port ${PORT}`);
    console.log(`📱 Webhook URL: http://localhost:${PORT}/webhook/whatsapp`);
    console.log(`🔧 Environment: ${process.env.NODE_ENV || 'development'}`);
  });
}

module.exports = {
  app,
  handleWhatsAppMessage,
  sendWhatsAppMessage
};

File: bots/whatsapp_bot.py
"""
WhatsApp Bot Handler (Python/FastAPI)
Handles incoming WhatsApp messages via Twilio
Processes claims and returns verification results
"""

import os
from fastapi import FastAPI, Form, Request
from fastapi.responses import Response
from twilio.twiml.messaging_response import MessagingResponse
from dotenv import load_dotenv
import sys

# Add parent directory to path for imports
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from modules.claim_extractor import analyze_message
from modules.hash_generator import generate_claim_hash
from modules.backend_client import submit_claim_to_backend, get_claim_result, claim_exists
from utils.formatter import (
    format_whatsapp_report,
    format_error_message,
    format_welcome_message,
    format_help_message
)

# Load environment variables
load_dotenv()

# Initialize FastAPI app
app = FastAPI(title="ThinkBaby WhatsApp Bot")

# Twilio configuration
TWILIO_ACCOUNT_SID = os.getenv("TWILIO_ACCOUNT_SID")
TWILIO_AUTH_TOKEN = os.getenv("TWILIO_AUTH_TOKEN")


async def handle_whatsapp_message(message_text: str, from_number: str) -> str:
    """
    Main message handler
    Process incoming WhatsApp messages
    
    Args:
        message_text: The message content
        from_number: Sender's phone number
        
    Returns:
        Response message
    """
    try:
        print(f"[WhatsApp] Message from {from_number}: {message_text}")
        
        # Handle commands
        lower_text = message_text.lower().strip()
        
        if lower_text in ['start', 'hi', 'hello']:
            return format_welcome_message()
        
        if lower_text == 'help':
            return format_help_message()
        
        # Analyze the message
        print("[WhatsApp] Analyzing message...")
        analysis = analyze_message(message_text)
        
        if not analysis.get("claims") or len(analysis["claims"]) == 0:
            return format_error_message("invalid_message")
        
        # Generate claim hash from first claim
        claim_text = analysis["claims"][0]
        claim_hash = generate_claim_hash(claim_text)
        print(f"[WhatsApp] Generated hash: {claim_hash}")
        
        # Check if claim exists in blockchain
        exists = claim_exists(claim_hash)
        
        if not exists:
            # Submit new claim to backend
            print("[WhatsApp] Submitting new claim to blockchain...")
            await submit_claim_to_backend(claim_hash, claim_text)
        
        # Get current result from blockchain
        print("[WhatsApp] Fetching blockchain result...")
        blockchain_result = get_claim_result(claim_hash)
        
        if not blockchain_result.get("success"):
            return format_error_message("api_down")
        
        # Format and return response
        report = format_whatsapp_report(
            analysis,
            blockchain_result.get("data", blockchain_result)
        )
        print("[WhatsApp] Report generated successfully")
        
        return report
    
    except Exception as e:
        print(f"[WhatsApp] Error handling message: {e}")
        return format_error_message("general")


@app.post("/webhook/whatsapp")
async def whatsapp_webhook(
    Body: str = Form(...),
    From: str = Form(...),
    request: Request = None
):
    """
    Webhook endpoint for incoming WhatsApp messages
    Twilio sends POST requests here
    """
    try:
        print(f"[Webhook] Received message: \"{Body}\" from {From}")
        
        # Process message
        response_message = await handle_whatsapp_message(Body, From)
        
        # Send response via Twilio
        twiml_response = MessagingResponse()
        twiml_response.message(response_message)
        
        return Response(content=str(twiml_response), media_type="application/xml")
    
    except Exception as e:
        print(f"[Webhook] Error: {e}")
        
        twiml_response = MessagingResponse()
        twiml_response.message(format_error_message("general"))
        
        return Response(content=str(twiml_response), media_type="application/xml")


@app.get("/health")
async def health_check():
    """Health check endpoint"""
    return {
        "status": "ok",
        "service": "WhatsApp Bot",
        "python": True
    }


@app.get("/")
async def root():
    """Root endpoint"""
    return {
        "service": "ThinkBaby WhatsApp Bot",
        "webhook": "/webhook/whatsapp",
        "health": "/health"
    }


# Run with: uvicorn whatsapp_bot:app --host 0.0.0.0 --port 3001 --reload
if __name__ == "__main__":
    import uvicorn
    port = int(os.getenv("WHATSAPP_BOT_PORT", 3001))
    print(f"✅ WhatsApp Bot server running on port {port}")
    print(f"📱 Webhook URL: http://localhost:{port}/webhook/whatsapp")
    uvicorn.run(app, host="0.0.0.0", port=port)
